elastic/logstash
# ECS Compatibility - backport 7x

## What does this PR do?

This PR is a clean backport of #12305 to 7.x:

>  Implements a plugin `ecs_compatibility` option, whose default value is powered
>  by the pipeline-level setting `pipeline.ecs_compatibility`, in line with the
>  proposal in elastic/logstash#11623:
>  
>  In order to increase the confidence a user has when upgrading Logstash, this
>  implementation uses the deprecation logger to warn when `ecs_compatibility` is
>  used without an explicit directive.
>  
>  For now, as we continue to add ECS Compatibility Modes, an opting into a
>  specific ECS Compatibility mode at a pipeline level is considered a BETA
>  feature. All plugins using the [ECS Compatibility Support][] adapter will
>  use the setting correctly, but pipelines configured in this way do not
>  guarantee consistent behaviour across minor versions of Logstash or the
>  plugins it bundles (e.g., upgraded plugins that have newly-implemented an ECS
>  Compatibility mode will use the pipeline-level setting as a default, causing
>  them to potentially behave differently after the upgrade).
>
> [ECS Compatibility Support]: https://github.com/logstash-plugins/logstash-mixin-ecs_compatibility_support

## Why is it important/What is the impact to the user?

This is a part of an ongoing effort to add ECS Compatibility modes to Logstash plugins (elastic/logstash#11635), and to have them on-by-default in the next major release of Logstash (elastic/logstash#11623). The newly-introduced `pipeline.ecs_compatibility` allows a user to explicitly control the default behaviour of all plugins in a pipeline.

This allows a user on Logstash 7.last to "lock in" a specific behaviour in advance of upgrading to Logstash 8.

## Checklist


## How to test this PR locally

The only plugin with ECS Compatibility that is currently shipped is the [Elasticsearch Output](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/952), whose behaviour is changed to use ECS-compatibile index templates. Configure a pipeline that uses it, and run Logstash with `--pipeline.ecs_compatibility=v1` set (command line) or `pipeline.ecs_compatibility: v1` (config, either `logstash.yml` or under a specific pipeline in `pipelines.yml`). Observe ECS-compatible templates being sent to Elasticsearch at startup.
